### PR TITLE
Rewrite DiscordRPC in Rust for Plugins v2

### PR DESCRIPTION
**Pull request by catornot**
:D

I can actually understand the code here

so this uses the official sdk  since the rust one doesn't work for me
( the C# files can be removed and probably c++, the reset is for bindgen )

image

I am not adding invites and lobbies since discord is removing that soon

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
```
Comment (GeckoEidechse) [inline Cargo.toml:1]: While we don't have build instructions for current plugins either, maybe we should add them for this PR.

At least something basic like:
- Compile C code using X
- Link with Rust code by running Y
- etc

^^

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
Given that we only target x86_64 we can probably omit the `aarch64` folder from the Discord game sdk ^^

**Comment by GeckoEidechse** [comment]
> I am not adding invites and lobbies since discord is removing that soon

Huh, didn't even know. We'll have to figure something out once we get to implementing our own party system, I guess ._.

File: Cargo.toml (hunk 1 above)
Comment (catornot) [reply Cargo.toml:1]: Yeah, I will add that soon

**Comment by GeckoEidechse** [comment]
CI fails cause it still assumes old build chain. Not an issue as I can fix it once I have build instructions ^^

File: Cargo.toml (hunk 1 above)
Comment (catornot) [reply Cargo.toml:1]: CI is done
just not sure why it doesn't upload artifacts ¯\_(ツ)_/¯

**Comment by GeckoEidechse** [comment]
With 8fabb123832947c21712f3ff2e1b54057dfe9a76 CI now passes and artifacts are also being uploaded :D

**Comment by GeckoEidechse** [comment]
Maybe merge main into this PR, so that the release CI file is also on this branch as well ^^
Then we can adjust it for new codebase.

Note that compared to the other two PRs for launcher and mods, I'm keeping this one a bit on the backburner as it depends on the other two anyway ^^

**Comment by catornot** [comment]
Done.

Also I will probably try another [crate](https://crates.io/crates/discord-sdk) for discord sdk that I found which is 100% rust :)

File: manifest/Resource.rc [hunk 2]
```diff
@@ -92,5 +92,4 @@ END
 
 
 /////////////////////////////////////////////////////////////////////////////
-#endif    // not APSTUDIO_INVOKED
-
+#endif    // not APSTUDIO_INVOKED
```
Comment (GeckoEidechse) [inline manifest/Resource.rc]: Any reason for newline removal

(bikeshedding is real btw)

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    user::User,
+    wheel::UserState,
+    wheel::Wheel,
+    Discord, DiscordApp, Subscriptions,
+};
+use rrplug::prelude::*;
+use rrplug::wrappers::presence::{GamePresence, GameStateEnum};
+use tokio::runtime::Runtime;
+
+const APP_ID: i64 = 941428101429231617;
```
Comment (GeckoEidechse) [inline src/lib.rs:18]: `APP_ID` is id for our application/discord-integration, right?

Value wasn't explained in old code either ^^"

**Review by GeckoEidechse — COMMENTED** [review]
Just a quick skim. A few more comments would be nice to help with readability ^^

File: manifest/Resource.rc (hunk 2 above)
Comment (catornot) [reply manifest/Resource.rc]: idk how that happened

File: src/lib.rs (hunk 3 above)
Comment (catornot) [reply src/lib.rs:18]: yes.
I took it from the c++ version

Comment (catornot) [reply src/lib.rs:18]: I believe @emma-miler owns that discord app

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+name = "northstar-discord-rpc"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rrplug = { git = "https://github.com/catornot/rrplug.git", features=["presence"] }
+# rrplug = { path = "../rrplug", features=["full"] }
```
Comment (GeckoEidechse) [inline Cargo.toml]: leftover commented out package

Comment (catornot) [reply Cargo.toml]: only nitpicks at this point XD

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "northstar-discord-rpc"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rrplug = { git = "https://github.com/catornot/rrplug.git", features=["presence"] }
```
Comment (pg9182) [inline Cargo.toml]: Pin rrplug, please.

**Review by pg9182 — CHANGES_REQUESTED** [review]
Looks good overall, will test soon.

**Review by pg9182 — CHANGES_REQUESTED** [review]
Crashes on https://github.com/catornot/rrplug/blob/461aac3e2a3ce112aa0c29143dceac42fe383dc1/src/bindings/unwraped.rs#L182.

```rust
            sq_schedule_call_external: value.__sq_schedule_call_external.unwrap(),
```

**Comment by catornot** [comment]
that's not supposed to happen o_o
issue with plugins v2?
(works for me btw)

File: Cargo.toml (hunk 5 above)
Comment (catornot) [reply Cargo.toml]: like to a commit? or actual version? 
(I will release rrplug v2 either this weekend or when 1.13 releases)

Comment (GeckoEidechse) [reply Cargo.toml]: Version is usually the way to go. I'd just release rrplug v2 now already tbh. Plan is to release 1.13 before 24th anyway, and that's like 2-3 days from now ^^

**Review by GeckoEidechse — CHANGES_REQUESTED** [review]
With launcher from https://github.com/R2Northstar/NorthstarLauncher/actions/runs/4748335619 and mods at https://github.com/R2Northstar/NorthstarMods/commit/f6716069a26dde85866741cc066f380ddf442f7d

it crashes for me on every launch as long as `northstar_discord_rpc.dll` from https://github.com/R2Northstar/NorthstarDiscordRPC/actions/runs/4647927747 is in `plugins` folder. Removing `northstar_discord_rpc.dll` it launches fine again.

**Review by GeckoEidechse — COMMENTED** [review]
No more crashing with 9f263e0cb48af117693063c42c36e6c234b5ae3e it seems 👀

**Review by GeckoEidechse — APPROVED** [review]
Works in testing :D

**Review by pg9182 — APPROVED** [review]
Looks good. Doesn't crash. Nice to finally get rid of the Discord SDK (also, the rust one is by the same person who wrote the windows sdk downloader I'm going to use for northstar).

**Pull request merged**